# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/gc/Zone.h
# Commit: 8e609f97c413
# Full Hash: 8e609f97c413a12f7042dfe80647bdd34f0400d8
# Author: Jon Coppeard <jcoppeard@mozilla.com>
# Date: 2019-11-01 16:13:11
# Regressor Bug: 1587096
# File Overlap Count: 1
# Description:
#   Bug 1587096 - Part 1: Add JS FinalizationGroupObject and related data structures r=sfink
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D49943
# ==============================================================================

diff -r f273ec2ec0ae -r 8e609f97c413 js/src/gc/Zone.h
--- a/js/src/gc/Zone.h	Fri Nov 01 10:37:09 2019 +0000
+++ b/js/src/gc/Zone.h	Fri Nov 01 10:37:25 2019 +0000
@@ -48,6 +48,9 @@
 template <typename T>
 class ZoneCellIter;
 
+// A vector of FinalizationRecord objects, or CCWs to them.
+using FinalizationRecordVector = GCVector<HeapPtrObject, 1, ZoneAllocPolicy>;
+
 }  // namespace gc
 
 using StringWrapperMap =
@@ -629,6 +632,18 @@
   uint32_t detachedTypedObjects = 0;
 
  private:
+  // A map from finalization group targets to a list of finalization records
+  // representing groups that the target is registered with and their associated
+  // holdings.
+  using FinalizationRecordMap =
+      GCHashMap<js::HeapPtrObject, js::gc::FinalizationRecordVector,
+                js::MovableCellHasher<js::HeapPtrObject>, js::ZoneAllocPolicy>;
+  js::ZoneOrGCTaskData<FinalizationRecordMap> finalizationRecordMap_;
+
+  FinalizationRecordMap& finalizationRecordMap() {
+    return finalizationRecordMap_.ref();
+  }
+
   js::ZoneOrGCTaskData<js::jit::JitZone*> jitZone_;
 
   js::MainThreadData<bool> gcScheduled_;